feat(schema): a new schema file for the new application system

abhiamishra opened the pull request:
a new schema file that combines the old schema with the new connections needed for application system

adds to #96

abhiamishra committed:
feat(schema): a new schema file for the new application system

a new schema file that combines the old schema with the new connections needed for application
system

adds to #96

abhiamishra commented:
Added tables from application system schema created in #96 and relations - review needed to ensure the relations and tables are of the correct format.

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,186 @@
+    typeformId          String
+    typeformName        String
+
+    user_app UserAppJunction @relation(fields: [formmId], references: [typeformId])
```
Remove this because Portal does not keep track of response data for Typeform application. There should be a separate Application object for the new app system.

abhiamishra replied:
Ok - for right now, ive replaced that entirely with a new Application object

DedsecKnight replied:
Might be better to keep both `TypeformApplication` and `Application` until we're 100% sure that the new app system is functional (just in case if something goes wrong, we can revert back to `TypeformApplication`)

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,186 @@
+    @@map("typeform_applications")
+}
+
+model UserAppJunction {
```
I don't think this model is setup correctly, assuming that this acts as a junction table between Application and Profile object. Refer to this guide to make sure things are setup correctly: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations.

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,186 @@
+    id                  String  @id @default(auto()) @map("_id") @db.ObjectId
+    deptName            String
+
+    officer Officer @relation(fields: [deptId], references: [id])
```
Based on schema, `Department` and `Officer` have many-to-many relationship, meaning that this should go to an intermediary table instead of the `Officer` model directly.

abhiamishra replied:
you are right - i mistakingly saw that as one to many. fixing it!

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,186 @@
+    user_app UserAppJunction @relation(fields: [deptId], references: [id])
+}
+
+model Officer {
```
I don't think this is set up correctly for 1-on-1 relationship between `Officer` and `Profile`. Refer to this to make sure that it is setup correctly: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/one-to-one-relations

abhiamishra replied:
yup - fixed on my end.

abhiamishra commented:
fixed issues by @DedsecKnight - will review it once more tomorrow and push

abhiamishra committed:
perf(improving the m-to-n relations and fixing other inefficiencies in the schema file): fix schema

Working on Mike's suggested changes on schema file and fixing the new schema in terms of Prisma's
format

works on #96

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+    user                User    @relation("UserLink", fields: [userId], references: [id])
+
+    formId              String 
+    form                Form    @relation("AppLink",  fields: [formId], references: [id]) 
```
> form Form @relation("AppLink", fields: [formId], references: [id])

Is there supposed to be a `Form` model?

abhiamishra replied:
sorry - that should be Application

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+    form                Form    @relation("AppLink",  fields: [formId], references: [id]) 
+
+    deptId              String
+    dept                Dept    @relation("DeptLink", fields: [deptId], references: [id])
```
I believe `Dept` should be `Department` instead, since that's the model name.

abhiamishra replied:
correct

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+    createdAt           DateTime
+    expireDate          DateTime
+    type                String?
+    role                String?
```
Is this `role` gonna be the same as the model `Role`? If so, might need to set up a relationship.

abhiamishra replied:
I was actually going to ask you about that. 

In the doc, it is defined as Role[] for the application but I'm unsure what that entails for an application?

abhiamishra replied:
done w the rest of changes - i just need your clarification on this please

DedsecKnight replied:
The `Role` object is for role-based access control. Currently, `officer` is the only role and it is used to ensure that only officers will have access to administrative features such as create events, applications, and Vanity link, and the plan is so that we can add another role like `sponsor` since there are plans to create features for sponsors to download resumes. Since it seems like applications are managed by division, perhaps it would be better to use `division` (or `department`) to control who has write permission to the application.

abhiamishra replied:
what is the difference between Role and RolesOnUser?

I'm thinking to just connect to Role with a 1-to-1

As per your division, I'll connect it to Division with a 1-to-1

What do you think?

DedsecKnight replied:
`Role` and `User` has a m-to-n relationship, in case a user can assume multiple roles, so `RolesOnUser` acts as a junction table. As for the division, I believe at some point there were plans to create a general application as there were overlapping questions in the application for cohort programs, so I would lean towards m-to-n relationship between `Application` and `Division` for now, unless there is a reason not to do so.

abhiamishra replied:
@DedsecKnight - can i get a resolve on this?

DedsecKnight replied:
>  I believe at some point there were plans to create a general application as there were overlapping questions in the application for cohort programs.

If this is not likely going to happen, feel free to resolve this conversation. Otherwise, I think we should maintain m-to-n relationship between `Application` and `Division` so that multiple divisions can own an application.

abhiamishra replied:
ok I will remove the role? variable since I have connected applications to a division that itself connects to officers so automatically, we already have the officers who can edit that specific application.

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+    profile       Profile?
+    roles         RolesOnUser[]
+
+    applications    UserAppJunction[]   @relation("UserLink")
```
Feels like this would be better if hooked up with the `Profile` model instead of `User`, as the `User` model in this schema is more for the purpose of compatibility with NextAuth, which is what we use for authentication, while `Profile` is more of a source of truth of user' data for member portal.

abhiamishra replied:
gotcha

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+    id                  String  @id @default(auto()) @map("_id") @db.ObjectId
+    deptName            String
+
+    officers            Officer[]
```
Based on [this part]() of the documentation, would we do something like this for officer-department? 
> officerIDs String[]
    officers    Officer[] @relation(fields: [officerIDs], references: [id])

abhiamishra replied:
yeah so in prisma, there are also implicit m-to-n relations where you just put it like that so instead of having a junction table, prisma takes care of building that beneath it.

that's why beneath it, i provided a junction table in case it is needed but the connection between officers and departments is a simple m-to-n so the implicit case should work here

DedsecKnight replied:
Based on the documentation, it seems like the implicit m-to-n relations is for SQL. For MongoDB, it seems like we have to do [this](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#mongodb).

abhiamishra replied:
you are correct - was looking at the wrong thing

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+    deptId              String
+
+    profile             Profile?
+    departments         Department[]
```
Same question as with `Officer` model

abhiamishra replied:
Wait I'm confused - can you expand on what the issue is here?

abhiamishra replied:
Look at my answer re: https://github.com/acmutd/portal-next/pull/113#discussion_r1064195316

DedsecKnight replied:
https://github.com/acmutd/portal-next/pull/113#discussion_r1065066370

DedsecKnight commented on apps/portal/prisma/new-schema.prisma:
```diff
@@ -0,0 +1,229 @@
+
+model Officer {
+    id                  String  @id @default(auto()) @map("_id") @db.ObjectId
+    deptId              String
```
Shouldn't this be `String[]`, since an officer can be a part of multiple division?

abhiamishra replied:
That actually needs to be removed by me - since the departments variable which is an array of Department takes care of that

DedsecKnight commented:
nit: might wanna change `Department` to `Division` so that anyone who later on maintain this project will not get confused.

abhiamishra commented:
I also have renamed the UserAppJunction as FilledApplication - makes more sense since Application is a model that will house the template application created and then FilledApplication are the applications with different information.

abhiamishra committed:
refactor(new-schema.prisma): changing how relations are done to be compliant with mongoDB

fixing syntax of relations to ensure it is compliant to how MongoDB runs and additionally, changing
some relations in some models

works on #96

abhiamishra committed:
perf(new-schema.prisma): removing the role? variable

removing the role? variable

works on #96

DedsecKnight committed:
Merge branch 'next' into newschema

DedsecKnight committed:
Merge branch 'next' into newschema

DedsecKnight committed:
Merge branch 'next' into newschema

DedsecKnight committed:
Merge branch 'next' into newschema

DedsecKnight committed:
Merge branch 'next' into newschema

DedsecKnight committed:
Merge branch 'next' into newschema

DedsecKnight committed:
refactor: update new-schema.prisma to match with schema.prisma

abhiamishra committed:
fixing new schema

abhiamishra committed:
committing again for fix

abhiamishra committed:
Pushing new re-factored branch

abhiamishra committed:
Importing changes into schema.prisma

DedsecKnight committed:
fixed build error
